### DISCUSSION TITLE
Help typing for CustomizedPage

### DISCUSSION THREAD

**Discussion by vunhatchuong** [uriyyo/fastapi-pagination#1159] [Q&A]
```python
PageOptionalParams =  CustomizedPage[Page, UseOptionalParams()]
@router.get(
    "/",
    dependencies=[Depends(pagination_ctx(PageOptionalParams[UserResponse]))])
async def get_all(db: AsyncSession = Depends(get_db) -> ResponseBase[PageOptionalParams[UserResponse]]:

    err, users = await user_crud.get_all(db)

    return create_response(error=err, data=users)
```
Error:
- Pyright: Type "Page[Unknown]" is already specialized (reportInvalidTypeArguments) **_on line `dependencies`_**
- Mypy: Variable "routers.user.PageOptionalParams" is not valid as a type **_on return type hint_**
- Pyright: Type "Page[Unknown]" is already specialized (reportInvalidTypeArguments) **_on return type hint_**

**Comment by uriyyo** [comment]
For mypy fix should be:
```py
T = TypeVar("T")
PageOptionalParams: TypeAlias = CustomizedPage[Page[T], UseOptionalParams()]

@router.get(
    "/",
    dependencies=[Depends(pagination_ctx(PageOptionalParams[UserResponse]))])
async def get_all(db: AsyncSession = Depends(get_db) -> ResponseBase[PageOptionalParams[UserResponse]]:

    err, users = await user_crud.get_all(db)

    return create_response(error=err, data=users)
```

**Reply by vunhatchuong** [reply]
Still some error with this
-  Invalid type alias: expression is not a valid type **_on line PageOptional Params: TypeAlias_ while using typing.TypeAlias**
-  Argument of type "PageOptionalParams[UserResponse]" cannot be assigned to parameter "page" of type "type[AbstractPage[Any]] | None" in function "pagination_ctx" Pyright (reportArgumentType)
 │        Type "type[Annotated]" is incompatible with type "type[AbstractPage[Any]] | None" 
 │          Type "type[Annotated]" is incompatible with type "type[AbstractPage[Any]]" 
 │          "type[type]" is incompatible with "type[None]"  **_on line dependencies_**

**Reply by uriyyo** [reply]
I guess it's `pyright` issue, I will take a look
